File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        n = len(matrix)
        if n==0:
            return False
        m = len(matrix[0])
        l=0
        r=n*m-1
        while(l<=r):
            mid = l+(r-l)//2
            x=mid//m
            y=mid%m
            if matrix[x][y]==target:
                return True
            elif matrix[x][y]>target:
                r=mid-1
            else:
                l=mid+1
        return False

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int n = matrix.length;
        if (n == 0) return false; 
        int m = matrix[0].length;
        int l = 0, r = n * m - 1;
        while (l <= r) { 
            int mid = l + (r - l) / 2;
            int x = mid / m;
            int y = mid % m;
            if (matrix[x][y] == target) {
                return true;
            } else if (matrix[x][y] < target) {
                l = mid + 1;
            } else {
                r = mid - 1;
            }
        }
        return false;
    }
}
